### PR TITLE
writen-report

### PR DESCRIPTION
**Pull request by wangwenju269**
## **User description**
**Features**

- Combine the open source large language model and the Metagpt framework to achieve intelligent writing. In order to verify the feasibility of the technical route, "accident reports" are used as the experimental subjects, and will be expanded to paper writing or project bidding (document level) writing.（将开源大语言模型和 Metagpt框架相互结合，实现智能写作工作。为验证技术路线的可行性，以“事故报告”做实验对象，后续将拓展到论文写作，或者项目招投标(文档级)写作上)

**Feature Docs**

 -  This function is document-level writing, which imitates human writing thinking. You write a first draft first, and then submit it to the teacher for evaluation and feedback. Later, based on the feedback, the first draft was revised and improved.
(该功能是文档级别写作，模仿人类的写作思维，先写出初稿，后交给老师评估，得到反馈意见。后根据反馈意见，进行对初稿修订改进)

**Influence**

- Liberate the writing ability of public officials. （解放公职人员写作能力）
-Implement generation of long document levels（实现长文档级别的生成）
**Result**

Generate papers, bidding documents, official documents, reports, etc.（生成论文，招投标文件，公文，报告等）
例如：
输入：
[“5.18”金达花园三期项目坍塌事故调查报告.json](https://github.com/geekan/MetaGPT/files/15218693/5.18.json)

输出：
(https://github.com/geekan/MetaGPT/files/15218678/5.18._metagpt.txt)

**Other**

___

## **Type**
enhancement

___

## **Description**
- Introduced a comprehensive system for generating, evaluating, and refining accident reports using the MetaGPT framework.
- Developed a new `RewriteReport` class with methods for planning, executing tasks, and handling tools for intelligent report creation.
- Added test scripts and actions for evaluating and refining the generated reports.
- Updated model configuration to "qwen-14B-Chat" for enhanced performance.
- Provided detailed guidance and structured task planning for different sections of an accident report.

___

## **Changes walkthrough**
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement
</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>core_report.py</strong><dd><code>Implement RewriteReport Class for Accident Report Generation</code></dd></summary>
<hr>
      
tests/write_roport/core_report.py

<li>Introduced a new class <code>RewriteReport</code> to handle the process of writing, <br>evaluating, and refining accident reports.<br> <li> Implemented methods for planning, acting on tasks, and handling tool <br>recommendations.<br> <li> Added functionality to write initial drafts, evaluate them, and refine <br>based on feedback.<br> <li> Integrated with the MetaGPT framework for intelligent document <br>creation.<br>

</details>
    

  </td>
  <td><a href="https://github.com/geekan/MetaGPT/pull/1235/files#diff-58fd511b722bbc7aa6edf10c79d3a8167479339cc3edbe0b8de1712a1effe9c9">+151/-0</a>&nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>write_evaluator_refine.py</strong><dd><code>Define Actions for Report Writing, Evaluation, and Refinement</code></dd></summary>
<hr>
      
tests/write_roport/write_evaluator_refine.py

<li>Defined actions for writing analysis, evaluation, and refinement of <br>reports.<br> <li> Utilized structured prompts and evaluation criteria for generating and <br>refining text.<br>

</details>
    

  </td>
  <td><a href="https://github.com/geekan/MetaGPT/pull/1235/files#diff-5935cbbff8c72554c3c7e4ca42bb44fe44fc6fbe01a6fe69252530c947fedd30">+110/-0</a>&nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>write_report_planner.py</strong><dd><code>Implement Planner for Structured Report Writing Tasks</code>&nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
tests/write_roport/write_report_planner.py

<li>Implemented a planner for report writing tasks based on user <br>requirements and SOP.<br> <li> Added functionality to update and manage plans for report writing.<br>

</details>
    

  </td>
  <td><a href="https://github.com/geekan/MetaGPT/pull/1235/files#diff-4851c1038df3ec2feb2adf63d065e0d22117fc85aa821a3977cb8f2a28b4e795">+102/-0</a>&nbsp; </td>

</tr>                    
</table></td></tr><tr><td><strong>Tests
</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>test_write_report.py</strong><dd><code>Test Script for RewriteReport Functionality</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
tests/write_roport/test_write_report.py

<li>Created a test script to instantiate and run the <code>RewriteReport</code> class <br>with specified parameters.<br>

</details>
    

  </td>
  <td><a href="https://github.com/geekan/MetaGPT/pull/1235/files#diff-cfddd6e17c1f5346514fdc96442312a035b4add9c769a97472ac65b366873b14">+13/-0</a>&nbsp; &nbsp; </td>

</tr>                    
</table></td></tr><tr><td><strong>Configuration changes
</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>config2.yaml</strong><dd><code>Update Model Configuration in YAML</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
config/config2.yaml

<li>Updated the model configuration from "gpt-4-turbo" to "qwen-14B-Chat".<br> <br>

</details>
    

  </td>
  <td><a href="https://github.com/geekan/MetaGPT/pull/1235/files#diff-f5014539003fa8de0547e0af4e482bf02b19072fdd5d3991671e59de67de708a">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>sop.json</strong><dd><code>Structured Order of Tasks for Report Writing in JSON</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
tests/data/report_writer/sop.json

<li>Defined a structured order of tasks for writing an accident report <br>based on user-provided information.<br>

</details>
    

  </td>
  <td><a href="https://github.com/geekan/MetaGPT/pull/1235/files#diff-312bd99b2755995ba3c616bb79f8e99fa09fa0d2340fd68eee9a5db52d06aa04">+38/-0</a>&nbsp; &nbsp; </td>

</tr>                    
</table></td></tr><tr><td><strong>Documentation
</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>domain.yaml</strong><dd><code>Define Domain-Specific Guidance for Report Writing</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
tests/data/report_writer/domain.yaml

<li>Added detailed descriptions and guidance for writing various sections <br>of an accident report.<br>

</details>
    

  </td>
  <td><a href="https://github.com/geekan/MetaGPT/pull/1235/files#diff-db10ba9df07e91d8b172b74c4a46e72c629e765c7426eaf0a2603ec7b58470e6">+98/-0</a>&nbsp; &nbsp; </td>

</tr>                    
</table></td></tr></tr></tbody></table>

___

> ✨ **PR-Agent usage**:
>Comment `/help` on the PR to get a list of all available PR-Agent tools and their descriptions

### CODE REVIEW

**Comment by wangwenju269** [comment]
按照返回意见，重新修改代码，清晰。对 原 Metagpt 结构 未作任何改变， 复用源基层代码， 模仿datainterpreter 类进行改造，实现与源代码 高耦合。

**Comment by geekan** [comment]
Great! Let's review

**Comment by wangwenju269** [comment]
I developed an official document writing artifact, which can be expanded to the fields of paper writing, official document generation, bidding documents and other fields according to user needs. In the future, tools such as networking and literature retrieval will continue to be integrated. The code repository is at: METAGPT\tests\metagpt\ext\write_report\report.py
(我开发一款公文写作神器， 根据用户的需求可以拓展到论文写作，公文生成，招投标文档等领域。后续会继续集成 联网，文献检索等工具。代码仓库在： METAGPT\tests\metagpt\ext\write_report\report.py)

**Comment by shenchucheng** [comment]
@wangwenju269 哈喽，有一些问题先请问下
1. 为什么代码是写在tests目录下？
2. 我看大部分的提示词都是中文，是必须使用中文提示词，还是什么原因呢？

**Comment by wangwenju269** [comment]
您好：

答复如下

       问题1 ： 放在 test 路径下的目的：自己开发的项目代码脚本 和 源metagpt框架 能更好解耦，本项目是复用 metagpt 现有的类和方法，代码做到高内聚，低耦合。

       问题2： 提示词是可以拓展的，已将提示词单独放在一个文件里。后续更换提示词文件，实现实现招投标、论文写作等应用场景。 提示词是可插拔替换的，根据具体业务场景，更换提示词

      







祝好！







在 2024-05-17 11:06:18，"shenchucheng" ***@***.***> 写道：

@wangwenju269 哈喽，有一些问题先请问下

为什么代码是写在tests目录下？
我看大部分的提示词都是中文，是必须使用中文提示词，还是什么原因呢？

**Comment by wangwenju269** [comment]
有修改意见，请及时反馈给我，谢谢

**Comment by garylin2099** [comment]
> 您好： 答复如下 问题1 ： 放在 test 路径下的目的：自己开发的项目代码脚本 和 源metagpt框架 能更好解耦，本项目是复用 metagpt 现有的类和方法，代码做到高内聚，低耦合。 问题2： 提示词是可以拓展的，已将提示词单独放在一个文件里。后续更换提示词文件，实现实现招投标、论文写作等应用场景。 提示词是可插拔替换的，根据具体业务场景，更换提示词 祝好！ 在 2024-05-17 11:06:18，"shenchucheng" ***@***.***> 写道： @wangwenju269 哈喽，有一些问题先请问下 为什么代码是写在tests目录下？ 我看大部分的提示词都是中文，是必须使用中文提示词，还是什么原因呢？ — Reply to this email directly, view it on GitHub, or unsubscribe. You are receiving this because you were mentioned.Message ID: ***@***.***>

metagpt下有ext文件夹（external, extension之意），可将代码放入metagpt/ext下，参考stanford_town的放置方法，可与metagpt核心代码相分开

File: tests/metagpt/ext/write_report/core_report.py [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+from metagpt.tools.tool_recommend import BM25ToolRecommender, ToolRecommender
+from tests.metagpt.ext.write_report.write_evaluator_refine import (
+    EvaluatorReport,
+    RefineReport,
+    WriteAnalysisReport,
+)
+from tests.metagpt.ext.write_report.write_report_planner import WritePlanner
+
+
+class RewriteReport(Role):
```
Comment (garylin2099) [inline tests/metagpt/ext/write_report/core_report.py:21]: 应为ReportRewriter，RewriteReport是动作的意思

File: config/config2.yaml [hunk 2]
```diff
@@ -3,6 +3,6 @@
 # Config Docs: https://docs.deepwisdom.ai/main/en/guide/get_started/configuration.html
 llm:
   api_type: "openai"  # or azure / ollama / groq etc.
-  model: "gpt-4-turbo"  # or gpt-3.5-turbo
-  base_url: "https://api.openai.com/v1"  # or forward url / other llm url
-  api_key: "YOUR_API_KEY"
\ No newline at end of file
+  model: "qwen-14B-Chat"  # or gpt-3.5-turbo
```
Comment (garylin2099) [inline config/config2.yaml:6]: 这个改动不应该提交

File: tests/metagpt/ext/write_report/report.py [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+
+import fire
+
+from tests.metagpt.ext.write_report.core_report import RewriteReport
+
+
+async def main(auto_run: bool = True):
+    requirement = "写一份事故报告"
+    di = RewriteReport(auto_run=auto_run, human_design_sop=True, use_evaluator=True)
+    await di.run(f"{requirement}", upload_file="E:\workspace\MetaGPT\data\“5.18”金达花园三期项目坍塌事故调查报告.json")
```
Comment (garylin2099) [inline tests/metagpt/ext/write_report/report.py:17]: 建议去掉具体的路径名和文件名

File: tests/metagpt/ext/write_report/write_evaluator_refine.py [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+"""
+
+REFINE = """请根据 `user` 的具体任务要求（Plan Status）以及以下的反馈建议，对 `assistant` 之前的回应进行深入审查，并在需要的地方进行适当的改进。
+用户反馈建议：{suggestion}
+
+在修订时，请保持与 `assistant` 原始回应的格式和风格一致，并确保输出的内容充分体现了用户的具体任务要求。
+"""
+
+
+def process_message(messages: Union[str, Message, list[dict], list[Message], list[str]]) -> list[dict]:
```
Comment (garylin2099) [inline tests/metagpt/ext/write_report/write_evaluator_refine.py:55]: 可使用llm.format_msg，见具体provider文件

File: tests/metagpt/ext/write_report/write_evaluator_refine.py [hunk 5]
```diff
@@ -0,0 +1,170 @@
+import json
+from typing import Union
+
+from metagpt.actions import Action
+from metagpt.schema import Message
+from metagpt.utils.common import CodeParser
+
+STRUCTUAL_PROMPT = """
```
Comment (garylin2099) [inline tests/metagpt/ext/write_report/write_evaluator_refine.py:8]: 可跟write_analysis_code.py保持一致，将prompt放入单独文件

File: tests/metagpt/ext/write_report/write_report_planner.py [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+
+    async def run(self, context: list[Message], max_tasks: int = 7, human_design_sop=True) -> str:
+        if not human_design_sop:
+            self.PROMPT_TEMPLATE = f"{self.PROMPT_TEMPLATE}\n\n{self.CONSTRAINTS}"
+            task_type_desc = "\n".join([f"- **{v['name']}**: {v['desc']}" for k, v in usertasktype.items()])
+            prompt = self.PROMPT_TEMPLATE.format(
+                context="\n".join([str(ct) for ct in context]), max_tasks=max_tasks, task_type_desc=task_type_desc
+            )
+            rsp = await self._aask(prompt)
+            rsp = CodeParser.parse_code(block=None, text=rsp)
```
Comment (garylin2099) [inline tests/metagpt/ext/write_report/write_report_planner.py:54]: 可改写PROMPT_TEMPLATE，这段使用super().run

File: tests/metagpt/ext/write_report/write_report_planner.py [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+                logger.warning(error_msg)
+                self.working_memory.add(Message(content=error_msg, role="assistant", cause_by=WritePlan))
+                max_retries -= 1
+                continue
+            _, plan_confirmed = await self.ask_review(trigger=ReviewConst.TASK_REVIEW_TRIGGER)
+        update_plan_from_rsp(rsp=rsp, current_plan=self.plan)
+        # working_memory
+        self.working_memory.clear()
+
+    def get_plan_status(self) -> str:
```
Comment (garylin2099) [inline tests/metagpt/ext/write_report/write_report_planner.py:85]: 这个函数是否需要改写？PLAN_STATUS待填入的变量名可以和原PLAN_STATUS保持一致

File: tests/metagpt/ext/write_report/write_report_planner.py [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+            )
+            rsp = await self._aask(prompt)
+            rsp = CodeParser.parse_code(block=None, text=rsp)
+        else:
+            with open(TEST_DATA_PATH / "report_writer/sop.json", "r", encoding="utf-8") as file:
+                rsp = json.dumps(json.loads(file.read()), ensure_ascii=False)  # 读取人类设计 `sop` 流程
+        return rsp
+
+
+class WritePlanner(Planner):
```
Comment (garylin2099) [inline tests/metagpt/ext/write_report/write_report_planner.py:61]: WriteReportPlanner

**Review by garylin2099 — COMMENTED** [review]
代码虽然集成了MG内DataInterpreter, WritePlan, Planner等类，但仍不免需要改写大量函数，子类函数与父类函数有较多重复。这并非此PR的问题，是原来DI及DI相关类未留出足够的可配置项。若从此PR出发，可以再仔细考虑，哪些地方是可以直接使用原DI代码，哪些是必须要自定义，尽量减少些重复

File: tests/metagpt/ext/write_report/write_report_planner.py (hunk 6 above)
Comment (wangwenju269) [reply tests/metagpt/ext/write_report/write_report_planner.py:54]: 不大认可，super().run函数里有TaskType这个枚举类型。在不改变源代码接触上，我之前考虑重新设置UserTaskType，但还需要外层封装一下，因此将 UserTaskType 里的内容卸载外部 yaml 文件里。

File: tests/metagpt/ext/write_report/write_evaluator_refine.py (hunk 5 above)
Comment (wangwenju269) [reply tests/metagpt/ext/write_report/write_evaluator_refine.py:8]: 这只是提示词应用模板，且是通用的，没必要存在单独文件里。

File: tests/metagpt/ext/write_report/report.py (hunk 3 above)
Comment (wangwenju269) [reply tests/metagpt/ext/write_report/report.py:17]: 好的

File: config/config2.yaml (hunk 2 above)
Comment (wangwenju269) [reply config/config2.yaml:6]: 只是为了接开源模型测试代码是否通过

File: tests/metagpt/ext/write_report/write_report_planner.py (hunk 7 above)
Comment (wangwenju269) [reply tests/metagpt/ext/write_report/write_report_planner.py:85]: 是的，之前我想复用来着，后来取消 TaskType 这个类的使用，就不能直接访问其中的方法了，就简单重写了

File: tests/metagpt/ext/write_report/write_evaluator_refine.py (hunk 4 above)
Comment (wangwenju269) [reply tests/metagpt/ext/write_report/write_evaluator_refine.py:55]: METAgpt 版本代码 更新速度有点快，二次开发不一定能跟上。

**Comment by wangwenju269** [comment]
答复：文件放置在 test 目录下，方便更好测试代码是否独立执行成功。
      我已遵循评审大佬的要求和反馈建议，做出相应修改，清再次查收。
      祝好！

File: config/config2.yaml (hunk 2 above)
Comment (shenchucheng) [reply config/config2.yaml]: 建议恢复到原先的改动：
image
本地测试时，使用~/.metagpt/config2.yaml文件进行配置

**Review by shenchucheng — CHANGES_REQUESTED** [review]

**Pull request closed without merging**